### PR TITLE
lang: funcs: Add datetime.format function

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

## Tips:

* please read the style guide before submitting your patch:
[docs/style-guide.md](../docs/style-guide.md)

* commit message titles must be in the form:

```topic: Capitalized message with no trailing period```

or:

```topic, topic2: Capitalized message with no trailing period```

* golang code must be formatted according to the standard, please run:

```
make gofmt		# formats the entire project correctly
```

or format a single golang file correctly:

```
gofmt -w yourcode.go
```

* please rebase your patch against current git master:

```
git checkout master
git pull origin master
git checkout your-feature
git rebase master
git push your-remote your-feature
hub pull-request	# or submit with the github web ui
```

* after a patch review, please ping @purpleidea so we know to re-review:

```
# make changes based on reviews...
git add -p		# add new changes
git commit --amend	# combine with existing commit
git push your-remote your-feature -f
# now ping @purpleidea in the github PR since it doesn't notify us automatically
```

### CODE REVIEW

File: lang/funcs/core/datetime/format_func.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"fmt"
+	"time"
+
+	"github.com/purpleidea/mgmt/lang/funcs/simple"
+	"github.com/purpleidea/mgmt/lang/types"
+)
+
+func init() {
+	simple.ModuleRegister(ModuleName, "format", &types.FuncValue{
+		T: types.NewType("func(a int, b str) str"),
```
Comment (purpleidea) [inline lang/funcs/core/datetime/format_func.go]: Should this be str, int to match the style of printf?

File: lang/funcs/core/datetime/format_func.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		V: Format,
+	})
+}
+
+// Format returns returns a textual representation of the input time.
```
Comment (purpleidea) [inline lang/funcs/core/datetime/format_func.go]: reflow to 80 chars please.

File: lang/funcs/core/datetime/format_func.go (hunk 1 above)
Comment (roidelapluie) [reply lang/funcs/core/datetime/format_func.go]: No, as it is different from printf. I would prefer to see the "date" first, then the format. Just like if we implement other functions with date, I would like to always have the date first.

Imagine we implement    ` func (t Time) After(u Time) bool`, I would like it to be:
`func(t int, u int) bool`

**Comment by purpleidea** [comment]
Not reflowed properly.

Can you write a test case for this, since it's sort of silly.

**Comment by roidelapluie** [comment]
why is it sort of silly??

**Comment by purpleidea** [comment]
> why is it sort of silly??

Would be great to have a test so I didn't have to check manually.

**Comment by roidelapluie** [comment]
added a test

File: lang/funcs/core/datetime/format_func_test.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+func TestFormat(t *testing.T) {
+	inputVal := &types.IntValue{V: 1443158163}
+	inputFormat := &types.StrValue{V: "2006"}
+
+	val, err := Format([]types.Value{inputVal, inputFormat})
+	if err != nil {
+		t.Error(err)
+	}
+	if val.String() != `"2015"` {
+		t.Error(fmt.Errorf("invalid output, expected %s, got %s", `"2015"`, val.String()))
```
Comment (purpleidea) [inline lang/funcs/core/datetime/format_func_test.go]: t.Errorf insteadf

File: lang/funcs/core/datetime/format_func.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+func init() {
+	simple.ModuleRegister(ModuleName, "format", &types.FuncValue{
+		T: types.NewType("func(a int, b str) str"),
+		V: Format,
+	})
+}
+
+// Format returns returns a textual representation of the input time. The
```
Comment (purpleidea) [inline lang/funcs/core/datetime/format_func.go]: reflow to 80 chars please

File: lang/funcs/core/datetime/format_func.go (hunk 1 above)
Comment (purpleidea) [reply lang/funcs/core/datetime/format_func.go]: Okay fine.

**Comment by purpleidea** [comment]
LGTM, rebased to master and merged, thanks again!

(I noticed the intermittent go.uber.org/zap failure-- I imagine it's just a problem with uber's infrastructure being temporarily down.)

**Pull request closed without merging**